### DISCUSSION TITLE
Parellalization of Circuits

### DISCUSSION THREAD

**Discussion by shreyas-londhe** [Consensys/gnark#338] [Q&A] [answered]
I am running some `for` loops in my circuit. Is there a way to parallelize the circuit logic? i.e can I use all the CPU cores for the prover to get faster proving?

**Comment by gbotrel** [comment, marked as answer]
hey, so, short answer, the prover already uses all available cores. 
Longer version;
frontend.Compile(circuit) -> constraint system
prover.Prove(constraint system, witness) -> 1. solve constraint system 2. encode proof

Encode proof is highly parallelized, and there isn't much room for iimprovment here (other than GPU / FPGA based acceleration). 
The first part, solving the constraint system, is "naively" parallelized; constraints that are independent are solved independently, but some pre-processing and smarter grouping of this graph of constraints could improve things a bit. However, in most settings, solving the constraint system is < 5% of the total proving time 👍 (exception for very large circuits running on very large machines).

**Reply by shreyas-londhe** [reply]
Thanks for the clarification, even I checked using `top` and saw that all CPU cores were being used. 

Can you tell me what is the maximum number of constraints possible with Gnark and which AWS instance shall we prefer to get maximum speed possible for Prover?

**Reply by gbotrel** [reply]
Maximum number of constraints depend on the curve 2-adicity. For BN254, you can fit ~128M constraints in a circuit I believe. 
The AWS Hpc6a is quite good for large circuits like that:-) . AMD chips tend to perform a bit better w/ our code base than intel ones.

**Reply by shreyas-londhe** [reply]
Gotcha! Thank you so much for the help :)

**Comment by rsinha** [comment]
Is there a way (within tests) to run prover with only 1 core? I am looking to do some benchmarking.

**Reply by gbotrel** [reply]
you can use the GOMAXPROCS or the `--cpus` flag in `go test`. but benchmarks on 1 core may not be representative (the data structure / algorithm is design with multiple go-routines in mind, it will have an unnecessary overhead with one CPU only).
